x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/aggs/categorization/CategorizeTextAggregator.java

```diff
@@ -0,0 +1,208 @@
+            public void collect(int doc, long owningBucketOrd) throws IOException {
+                collectFromSource(doc, owningBucketOrd);
+            }
+
+            private void collectFromSource(int doc, long owningBucketOrd) throws IOException {
+                sourceLookup.setSegmentAndDocument(ctx, doc);
+                Iterator<String> itr = sourceLookup.extractRawValues(sourceFieldName).stream().map(obj -> {
+                    if (obj == null) {
+                        return null;
+                    }
+                    if (obj instanceof BytesRef) {
+                        return fieldType.valueForDisplay(obj).toString();
+                    }
+                    return obj.toString();
+                }).iterator();
+                while (itr.hasNext()) {
+                    TokenStream ts = analyzer.tokenStream(fieldType.name(), itr.next());
+                    processTokenStream(owningBucketOrd, ts, doc);
+                }
+            }
+
+            private void processTokenStream(long owningBucketOrd, TokenStream ts, int doc) throws IOException {
+                try {
+                    CharTermAttribute termAtt = ts.addAttribute(CharTermAttribute.class);
+                    ts.reset();
+                    ArrayList<BytesRef> tokens = new ArrayList<>();
+                    while (ts.incrementToken()) {
+                        tokens.add(new BytesRef(termAtt));
+                    }
+                    if (tokens.isEmpty()) {
+                        return;
+                    }
+                    categorizers = bigArrays.grow(categorizers, owningBucketOrd + 1);
+                    CategorizationTokenTree categorizer = categorizers.get(owningBucketOrd);
+                    if (categorizer == null) {
+                        categorizer = new CategorizationTokenTree(maxChildren, maxDepth, similarityThreshold);
+                        addRequestCircuitBreakerBytes(categorizer.ramBytesUsed());
+                        categorizers.set(owningBucketOrd, categorizer);
+                    }
+                    long previousSize = categorizer.ramBytesUsed();
+                    LogGroup lg = categorizer.parseLogLine(tokens.toArray(BytesRef[]::new), docCountProvider.getDocCount(doc));
+                    long newSize = categorizer.ramBytesUsed();
+                    if (newSize - previousSize > 0) {
+                        addRequestCircuitBreakerBytes(newSize - previousSize);
+                    }
+
+                    long bucketOrd = bucketOrds.add(owningBucketOrd, lg.getId());
+                    if (bucketOrd < 0) { // already seen
+                        bucketOrd = -1 - bucketOrd;
+                        collectExistingBucket(sub, doc, bucketOrd);
+                    } else {
+                        lg.bucketOrd = bucketOrd;
+                        collectBucket(sub, doc, bucketOrd);
+                    }
+                } finally {
+                    ts.close();
```

Comment:
Maybe let go of the token stream's contents before collecting the child buckets?

Reply:
I can do that, it makes sense here as we read all the tokens before collecting the children (unlike significant text)